From 528ae381b0d130cbd3fb9dcdbe3d99246890fca3 Mon Sep 17 00:00:00 2001
From: Rakesh Babu <rsaladi2@marvell.com>
Date: Fri, 4 Oct 2019 12:37:24 +0530
Subject: [PATCH 0295/1921] octeontx2-af: Get MSIXTR_BASE from fw_data

Get MSIX base address from firmware via fwdata structure
instead of from CSR as the later will cause issue when kernel crashes
and Kexec/Kdump loads secondary kernel and AF driver in the secondary
kernel reads IOVA written by primary.

Change-Id: I5a54638e2909e208537c68d4725f8f159201fc9d
Signed-off-by: Rakesh Babu <rsaladi2@marvell.com>
Reviewed-on: https://sj1git1.cavium.com/17057
Reviewed-by: Sunil Kovvuri Goutham <Sunil.Goutham@cavium.com>
Tested-by: Sunil Kovvuri Goutham <Sunil.Goutham@cavium.com>
[WK: The original patch got from Marvell sdk11.21.09]
Signed-off-by: Wenlin Kang <wenlin.kang@windriver.com>
---
 drivers/net/ethernet/marvell/octeontx2/af/rvu.c | 9 ++++++---
 drivers/net/ethernet/marvell/octeontx2/af/rvu.h | 3 ++-
 2 files changed, 8 insertions(+), 4 deletions(-)

diff --git a/drivers/net/ethernet/marvell/octeontx2/af/rvu.c b/drivers/net/ethernet/marvell/octeontx2/af/rvu.c
index 534c64c8ddff..4cfb4bed5bee 100644
--- a/drivers/net/ethernet/marvell/octeontx2/af/rvu.c
+++ b/drivers/net/ethernet/marvell/octeontx2/af/rvu.c
@@ -617,7 +617,10 @@ static int rvu_setup_msix_resources(struct rvu *rvu)
 	 */
 	cfg = rvu_read64(rvu, BLKADDR_RVUM, RVU_PRIV_CONST);
 	max_msix = cfg & 0xFFFFF;
-	phy_addr = rvu_read64(rvu, BLKADDR_RVUM, RVU_AF_MSIXTR_BASE);
+	if (rvu->fwdata && rvu->fwdata->msixtr_base)
+		phy_addr = rvu->fwdata->msixtr_base;
+	else
+		phy_addr = rvu_read64(rvu, BLKADDR_RVUM, RVU_AF_MSIXTR_BASE);
 	/* Register save */
 	rvu->msixtr_base_phy = phy_addr;
 	iova = dma_map_resource(rvu->dev, phy_addr,
@@ -904,6 +907,8 @@ static int rvu_setup_hw_resources(struct rvu *rvu)
 
 	mutex_init(&rvu->rsrc_lock);
 
+	rvu_fwdata_init(rvu);
+
 	err = rvu_setup_msix_resources(rvu);
 	if (err)
 		return err;
@@ -927,8 +932,6 @@ static int rvu_setup_hw_resources(struct rvu *rvu)
 		rvu_scan_block(rvu, block);
 	}
 
-	rvu_fwdata_init(rvu);
-
 	err = rvu_npc_init(rvu);
 	if (err)
 		goto fwdata_err;
diff --git a/drivers/net/ethernet/marvell/octeontx2/af/rvu.h b/drivers/net/ethernet/marvell/octeontx2/af/rvu.h
index 6cb284b8027c..b50b99cda443 100644
--- a/drivers/net/ethernet/marvell/octeontx2/af/rvu.h
+++ b/drivers/net/ethernet/marvell/octeontx2/af/rvu.h
@@ -392,7 +392,8 @@ struct rvu_fwdata {
 	u64 rclk;
 	u64 mcam_addr;
 	u64 mcam_sz;
-#define FWDATA_RESERVED_MEM 1024
+	u64 msixtr_base;
+#define FWDATA_RESERVED_MEM 1023
 	u64 reserved[FWDATA_RESERVED_MEM];
 	/* Do not add new fields below this line */
 #define CGX_MAX         4
-- 
2.31.1

